ESPS/general/src/vq/vqdes.1
.\" @(#)vqdes.1	3.8 4/7/97 ESI
.\" Copyright (c) 1987 Entropic Speech, Inc.; All rights reserved
.TH VQDES 1\-ESPS 4/7/97
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Speech, Inc.
.SH "NAME"
vqdes - design a vector quantization codebook from a FEA file
.SH "SYNOPSIS"
.B vqdes
[
.BI \-x " debug_level"
] [
.BI \-P " param_file"
] [
.BI \-h " histfile"
] [
.BI \-k " checkfile"
] [
.BI \-i 
] 
.I "input.fea output.fea"
.SH "DESCRIPTION"
.PP
.I Vqdes
takes as input a file
.I input.fea 
and obtains a training sequence of feature vectors, each consisting 
of a contiguous subset of elements from a FEA file field of type FLOAT
(use \fIfeafunc\fP(1\-\s-1ESPS\s+1) to transform types).  
.I Vqdes
designs a vector quantization codebook from the input training sequence
with codebook characteristics determined from information in the ESPS
parameter file (see ESPS PARAMETERS).  
.PP
If the \fB\-i\fR option is given, an initial codebook is read 
from a file determined by information in the ESPS parameter file.  
.PP
The final 
codebook and all converged codebooks of intermediate sizes 
are output as records in an ESPS FEA_VQ file 
.I output.fea.  
If "\-" is supplied in place of 
.I output.fea, 
then standard output is used.  
.PP
After every pass through the training sequence, the centroid-adjusted
codebook is output to a checkpoint file.  Only a single record is kept
in the checkpoint file (it is over-written each time).  
.PP
If the total memory required to store the feature vectors from 
.I input.fea 
is less than an internal limit of 
.I vqdes,
the codebook design algorithm will run in memory.  Otherwise, 
.I vqdes
will obtain pieces from 
.I input.fea
as necessary.  
.PP
For a detailed description of the VQ design algorithm, 
see
.I vqdesign
(3\-ESPS).  
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-x " debug_level \fR[2]\fP"
If 
.I "debug_level"
is positive,
.I vqdes
outputs debugging messages and other information about the codebook
design to a history file.  The messages proliferate as 
.I debug_level 
increases.  Levels up to 6 are supported currently.  For
information about what output results from each value of 
.I debug_level,
see the discussion of the parameter 
.I hdetail
in
.I vqdesign
(3\-ESPS).  
.IP
For the default level (2), a summary of the various design parameters
is output to the history file at the beginning of the run.  Then, the
following are written each time the codebook has converged at a given
size or after an empty cell has been filled:  current date and time,
current codebook size, size of training sequence for current
codebook, total number of clustering iterations so far, average
distortion of the current codebook, and the total number of empty
cells found so far.  Also, the average distortion with respect to the
current codebook is output each time the full training sequence is
encoded.  
.TP
.BI \-h " histfile \fR[vqdeshist]\fP"
Specifies the file to use for history and debugging output.  
.TP
.BI \-k " checkfile \fR[vqdes.chk]\fP"
Specifies the name of the checkpoint file written by 
.I vqdes.  
\.TP
.B \-i
Specifies that an initial codebook is to be used (otherwise a root
codebook is designed from the training sequence).  The filename 
and record number for the initial codebook are given respectively
by the parameters 
.I init_file
and
.I init_rec
in the ESPS parameter file.  The file must be of type FEA_VQ.  
Whether or not the initial codebook is clustered before splitting 
is determined by the parameter
.I init_behav
in the ESPS parameter file.  
.IP
The parameters
.I conv_ratio
and
.I dist_type
apply whether or not the \fB\-i\fP option is used.  If the parameter
.I cbk_type
is supplied (it's optional) it must be the same as the codebook type
of the initial codebook.  
.SH "ESPS PARAMETERS"
.PP
The following parameters are read from the parameter or common 
file:
.sp
.I "string fea_field"
.PP
This is the name of the field in \fIinput.fea\fP that contains 
the training vectors.  It must be of type FLOAT.  
.sp
.I "int fea_dim"
.PP
This is the dimension of the feature vectors and also the dimension 
of the VQ codewords.   
.I Fea_dim
must be no larger than the size of 
.I fea_field.
If there are more than 
.I fea_dim
elements in 
.I fea_field,
only the first 
.I fea_dim
are used from each record.  
.sp
.I "double conv_ratio"
.PP
Each clustering iteration of the codebook design algorithm is 
repeated until the fractional decrease in the overall average 
distortion between successive iterations falls below
.I conv_ratio.  
A reasonable initial value is .05
.sp
.I "int vq_size"
.PP
Specifies the size of the desired VQ codebook.
.sp
.I "string dist_type"
.PP
Specifies the type of distortion measure to use.  Currently supported
values are "MSE" (mean-square-error) and "MSE_LAR" (mean-square-error on log
area ratios, which assumes that the input vectors are reflection
coefficients).  
.sp
.I "string cbk_struct"
.PP
Specifies the search-structure of the desired codebook.  "FULL_SEARCH"
is the only value supported at this time.  
.sp
.I "string cbk_type"
.PP
Specifies a codebook type; the values currently supported are "MISC",
"RC_VQCBK", "RC_VCD_VQCBK", "RC_UNVCD_VQCBK", "LSF_VQCBK",
"LSF_VCD_VQCBK", and "LSF_UNVCD_VQCBK" (see <vq.h>).  Because this parameter
is stored as extra information in the codebook and does not affect
the codebook design, it is optional. The value "MISC" is assumed if 
the parameter is missing.  (However, if the parameter is present when
the \fB\-i\fP option is used, then it must match the value in the 
initial codebook.)
.sp
.I "string init_file"
.PP
Specifies the name of the file containing the initial codebook (only 
used with \fB\-i\fP option).  
.sp
.I "int init_rec"
.PP
Specifies the record number of the initial codebook in 
.I init_file
(only used with \fB\-i\fP option).  
.sp
.I "string init_behav"
.PP
If this parameter is "INIT_NOCLUSTER", the initial codebook is 
split right away.  If it is "INIT_CLUSTER", the training sequence 
is used to recluster the initial codebook (until convergence) before
splitting (only used with \fB\-i\fP option). 
.sp
.I "int max_iter"
Specifies the maximum number of iterations per level.  If this
parameter is not present, a default value of 500 is used.  
.PP
The values of parameters obtained from the parameter file are printed
if the environment variable ESPS_VERBOSE is 3 or greater.  The default
value is 3.
.SH ESPS COMMON
.PP
ESPS Common is read if
.I filename
in Common matches 
.I input. 
No values are written to ESPS Common. 
.PP
ESPS Common processing may be disabled by setting the environment variable
USE_ESPS_COMMON to "off".  The default ESPS Common file is .espscom 
in the user's home directory.  This may be overidden by setting
the environment variable ESPSCOM to the desired path.  User feedback of
Common processing is determined by the environment variable ESPS_VERBOSE,
with 0 causing no feedback and increasing levels causing increasingly
detailed feedback.  If ESPS_VERBOSE is not defined, a default value of 3 is
assumed.
.SH ESPS HEADERS
.PP
.I Vqdes
sets 
.I hd.fea\->type
to FEA_VQ, and it fills in the other header fields so 
as to define FEA records of that type. The command line is 
added as a comment, and 
.I input.fea
is added as a source file.  If there is an initial codebook
specified, its header is also added as a source.  
.I Vqdes
also creates and writes the following generic header items:
.sp
.I "design_size"
.PP
This is the number of codewords in the codebook.  
.sp
.I "codeword_dimen"
.PP
This is the dimension of each codeword.  
.sp
.I "fea_field"
.PP
This is the name of the field in \fIinput.fea\fP that was used
to form the training sequence.  
.PP
Note that both 
.I "design_size"
and
.I "codeword_dimen"
are redundant since 
the same values are stored in the FEA record itself.  However, 
they are also written in the header as they do determine the size
of the FEA file records.  
.SH "FUTURE CHANGES"
.PP
Additional values for the parameters
.I dist_type,
and
.I cbk_struct 
will be supported.  
.I Vqdes
will be modified so that standard input can be used in 
place of 
.I input.fea.
.SH "SEE ALSO"
.PP
.nf
vq(1\-ESPS), vqdesasc(1\-ESPS), vqasc(1\-ESPS), vqdesign(3\-ESPS), 
\fIfeafunc\fP(1\-\s-1ESPS\s+1), \fImergefea\fP(1\-\s-1ESPS\s+1), FEA_VQ(5\-ESPS), 
ESPS(5-ESPS), FEA(5-ESPS)
.fi
.SH "WARNINGS AND DIAGNOSTICS"
.PP
.I Vqdes 
will exit with a diagnostic if any of the following hold:
.IP
An initial codebook is used and its file name is the same as that of
the checkpoint file (this is for safety).
.IP
It can't read the input ESPS file headers or they are not of the right type.
.IP
The named field in the FEA file is not large enough.
.IP
There are insufficient records in the initial codebook file.
.IP
The named field in the FEA file is not of type FLOAT.
.IP
The dimension of the initial codebook doesn't match the dimension of
vectors in the training sequence.
.IP
An invalid distortion type was specified.  
.IP
Not enough memory could be allocated.
.IP
The parameter \fIcbk_type\fP doesn't match that of the initial
codebook when the \fB\-i\fP option is used.  
.PP
Warnings are issued if any of the following hold:
.IP
The FEA field size is larger than the specified dimension of training 
vectors.
.IP
The initial codebook type is not the same as that specified.
.IP
The initial codebook does not have a named fea_field that is the same
as that specified, or it does but it is different.  
.IP
Various warnings if characteristics of the initial codebook (\fB\-i\fP 
option) are inconsistent with the design parameters.  
.SH "BUGS"
.PP
None known.  
.SH "REFERENCES"
.nf
\fIvqdesign\fP(3\-\s-1ESPS\s+1), \fIvq\fP(1\-\s-1ESPS\s+1), \fIvqdesasc\fP(1\-\s-1ESPS\s+1), 
\fIvqasc\fP(1\-\s-1ESPS\s+1), \fIpplain\fP(1\-\s-1ESPS\s+1), \fIaddfea\fP(1\-\s-1ESPS\s+1).  
.fi
.SH "AUTHOR"
.PP
Manual page by John Shore; program by John Shore.  



ESPS/general/src/vq/vqdes.c
/*
 * This material contains unpublished, proprietary software of 
 * Entropic Research Laboratory, Inc. Any reproduction, distribution, 
 * or publication of this work must be authorized in writing by Entropic 
 * Research Laboratory, Inc., and must bear the notice: 
 *
 *    "Copyright (c) 1986-1990  Entropic Speech, Inc. 
 *    "Copyright (c) 1990-1996  Entropic Research Laboratory, Inc. 
 *                   All rights reserved"
 *
 * The copyright notice above does not evidence any actual or intended 
 * publication of this source code.     
 *
 * Written by:  John Shore
 * Checked by:
 * Revised by:
 *
 * Brief description:  This program is used to design a VQ codebook from 
 *                     FEA data
 *
 */

static char *sccs_id = "@(#)vqdes.c	3.10	1/20/97	ESI/ERL";
int  	    debug_level = 2;	    /*level of history detail*/

/*
 * include files
 */
#include <stdio.h>
#include <esps/esps.h>
#include <esps/fea.h>
#include <esps/vq.h>
/*
 * defines
 */
#define Fprintf (void)fprintf
#define Fflush (void)fflush
#define DEBUG(n) if (debug_level >= n) Fprintf
#define ERROR_EXIT(text) {(void) fprintf(stderr, "%s: %s - exiting\n", \
		ProgName, text); exit(1);}
#define SYNTAX USAGE ("vqdes [-x debug_level] [-P param_file] [-h histfile] [-k checkfile] [-i] input.fea output.fea")
#define FEA_BUFSIZE 1000000
#define MAX_ITER 500
/*
 * system functions and variables
 */
int getopt ();
double	atof();
double log();
long time();
int atoi(), strcmp(), fclose();
void rewind();
extern  optind;
extern	char *optarg;
char *mktemp(), *ctime(), *strcpy();
void rewind();
#ifndef DEC_ALPHA
char *calloc();
#endif
/*
 * external SPS functions
 */
int symtype();
long *add_genhd_l();
short *add_genhd_s();
float *add_genhd_f();
char *get_genhd();
char *add_genhd_c();
long get_fea_siz();
short get_fea_type();
char *get_cmd_line(), *getsym_s();
void symerr_exit(), write_header(), set_sd_type();
int getsym_i(), lin_search();
struct header *read_header();
struct fea_data *allo_fea_rec();
float **f_mat_alloc();
int rc_to_lar();
void vec_copy_f();
/*
 * global function declarations
 */
long rc_get_chunk();
double mse_lar_dist();
void s_param_err();
int checkpnt();
/*
 * global variable declarations
 */
char		    *ProgName = "vqdes";
char		    *cmd_line;		/*string for command line*/
char		    *histfile = "vqdeshist";	/*output ASCII file for history data*/
FILE		    *histrm;
char		    *input_ana = NULL;  /*input FEA file*/
FILE		    *fea_strm = NULL;
struct header	    *fea_ih;
struct fea_data	    *fea_rec;
char		    *fea_field = NULL;  /*name of FEA data field*/
long		    ndrec_ana;		/* num records in FEA file
					   or -1 if unknown */
int		    whole_fits;		/* flag for whether all FEA data fits
					   in feadata */
char		    *out_vq = NULL;	/*output FEA file for codebook*/
struct header	    *vq_oh;
FILE		    *vq_strm = NULL;
char		    *checkfile = "vqdes.chk";    /*checkpoint file*/
FILE		    *chkstrm;
struct header	    *chk_oh;
/*
 * main program
 */
main (argc, argv)
int argc;
char **argv;
{
/*
 * setup and initialization
 */
    char		*Version = "3.10";
    char		*Date = "1/20/97";
    char		*paramfile = NULL;  /*parameter file name*/
    char		*init_file = NULL;  /*file with initial codebook*/
    FILE		*init_strm;
    struct header	*init_ih;
    int			init_rec;	    /*record number of initial codebook*/
    int			init_behav;	    /*whether to cluster initial codebook*/
    int			c;		    /*for getopt return*/
    long		tloc;		    /*for local time*/
    long		i, j;		    /*loop variables*/
    long		fea_size;	    /*size of FEA field data*/
    short		vqfea_type;	    /*generic header item type in
					      initial codebook*/
    float		*fea_ptr;	    /* location of FEA field data
					       in fea_rec */
    float		**feadata;	    /*matrix of training FEA data*/
    long		*enc;		    /*final encode indices for data*/ 
    long		fea_len;	    /*numb. rows in feadata*/
    long		fea_dim;	    /*numb. columns in feadata*/
    long		c_rows;		    /*size of VQ codebook*/
    struct vqcbk	*cbk;		    /*the VQ codebook*/
    struct vqcbk	*icbk;		    /*an initial VQ codebook*/
    int			cbk_type = -1;    /*type of codebook (param file)*/
    double		(*distort)();	    /*routine to use for distortion*/
    int			vq_ret;		    /*return value from vqdesign*/    
    int			init_flag = 0;	    /*flag for -i option*/
    int			max_iter = MAX_ITER;/*max no. of iterations*/
/*
 *initialization
 */
    cmd_line = get_cmd_line(argc, argv); /*store copy of command line*/
/*
 * process command line options
 */

    while ((c = getopt (argc, argv, "x:P:h:k:i")) != EOF) {
	switch (c) {
	    case 'x': 
		debug_level = atoi (optarg);
		break;
	    case 'P':
		paramfile = optarg;
		break;
	    case 'h':
		histfile = optarg;
		break;
	    case 'k':
		checkfile = optarg;
		break;
	    case 'i':
		init_flag++;
		break;
	    default:
		SYNTAX;
	}
    }
/*
 * process file arguments
 */
    if (optind < argc) {
	input_ana = argv[optind++];
	if (strcmp (input_ana, "-") == 0) {
	    input_ana = "<stdin>"; 
	    /*For now, can't use stdin*/
	    ERROR_EXIT("can't use standard input for training sequence");
	}
	else
	    TRYOPEN (argv[0], input_ana, "r", fea_strm);
	}
    else {
	Fprintf(stderr, "vqdes: no input FEA file specified.\n");
	SYNTAX;
        }
    if (optind < argc) {
	out_vq = argv[optind++];
	if (strcmp (out_vq, "-") == 0)
	    out_vq = "<stdout>";
	else
	    TRYOPEN (argv[0], out_vq, "w", vq_strm);
	}
    else {
	Fprintf(stderr, "vqdes: no output file specified.\n");
	SYNTAX;
        }
    TRYOPEN(argv[0], histfile, "w", histrm);
/*
 * initial history and debug output
 */
    if (debug_level > 0) {
        tloc = time(0);
        Fprintf(histrm, "vqdes history output on %s", ctime(&tloc));
        Fprintf(histrm, "vqdes version %s, date %s\n", Version, Date);
        Fprintf(histrm, "command line:\n%s", cmd_line);
	Fprintf(histrm, "checkpoint file is %s\n", checkfile);
	Fflush(histrm);
    }
/*
 * read FEA header and allocate record
 */
    if ((fea_ih = read_header(fea_strm)) == NULL)      
	ERROR_EXIT("couldn't read input FEA file header");
    if (fea_ih->common.type != FT_FEA)
	ERROR_EXIT("input not FEA file");
    ndrec_ana = fea_ih->common.ndrec;
    fea_rec = allo_fea_rec(fea_ih);
/*
 * read parameters from parameter file and SPS common - require that
 * the filename in common match that of the input FEA file
 */
    (void) read_params(paramfile, SC_CHECK_FILE, input_ana);
    if(symtype("max_iter") != ST_UNDEF) 
	max_iter = getsym_i("max_iter");
    fea_dim = getsym_i("fea_dim");
    c_rows = getsym_i("vq_size");
    fea_field = getsym_s("fea_field");
    symerr_exit();  /*exit if any of the parameters were missing*/
    if (init_flag) {
	/*
         *get name and record for initial codebook -- make sure file is ok
	 */
	init_file = getsym_s("init_file");
	if (strcmp(checkfile, init_file) == 0)
	    ERROR_EXIT("Checkpoint file has same name as init_file.")
	init_rec = getsym_i("init_rec");
	TRYOPEN(argv[0], init_file, "r", init_strm);
	if ((init_ih = read_header(init_strm)) == NULL) 
	    ERROR_EXIT("couldn't read input FEA_VQ file header");
	if ((init_ih->common.type != FT_FEA) || (init_ih->hd.fea->fea_type != FEA_VQ))
	    ERROR_EXIT("initial codebook file not FEA or not FEA_VQ");
    }
    /*
     * cbk_type is optional
     */
    if (symtype("cbk_type") != ST_UNDEF) 
	if ((cbk_type = lin_search(vq_cbk_types, getsym_s("cbk_type"))) == -1)
	    ERROR_EXIT("invalid cbk_type in parameter file");
    /*
     *check to make sure fea_field exists, has right type, and is big
     * enough
     */
    fea_size = get_fea_siz(fea_field, fea_ih, (short *) NULL, (long **) NULL);
    if (fea_size == 0) {
	Fprintf(stderr, fea_field, 
	    "vqdes: field %s doesn't exist in input file\n", fea_field);
	exit(1);
    }
    if (fea_dim > fea_size) ERROR_EXIT("VQ_DIM larger than FEA field size");
    if (fea_dim != fea_size)  
	Fprintf(histrm, 
	  "vqdes: WARNING - size of FEA field greater than VQ_DIM\n");
    if (get_fea_type(fea_field, fea_ih) != FLOAT) 
	ERROR_EXIT("specified fea_field is not type FLOAT");
/*
 *create FEA header 
 */
    vq_oh = new_header(FT_FEA);
    if (init_vqfea_hd(vq_oh, c_rows, fea_dim) != 0) 
	ERROR_EXIT("error filling FEA header");
    (void) strcpy (vq_oh->common.prog, "vqdes");
    (void) strcpy (vq_oh->common.vers, Version);
    (void) strcpy (vq_oh->common.progdate, Date);
    add_source_file(vq_oh, input_ana, fea_ih);
    if (init_flag) add_source_file(vq_oh, init_file, init_ih);
    add_comment(vq_oh, cmd_line);
    /*
     *     *add generic header item for field name
     */
    (void) add_genhd_c("fea_field", fea_field, 0, vq_oh);
/*
 *Now that the header is complete, we can write it out
 */
    write_header(vq_oh, vq_strm);
    Fflush(vq_strm); /*just in case vqdes bombs :-)*/
/*
 *Allocate FEA_VQ record.
 */
    cbk = allo_vqfea_rec(vq_oh);
/*
 *Now finish processing the parameters and initializing the codebook
 */
    if (init_flag) {
	if ((init_behav = lin_search(vq_init, getsym_s("init_behav"))) == -1){
	    Fprintf(stderr, "vqdes: invalid init_behav in parameter file\n");
	    exit(1);
	}
	/* error and consistency checking
	 */
	if (init_ih->common.ndrec < init_rec && init_ih->common.ndrec != -1)
	    ERROR_EXIT("not enough records in initial codebook file");
	/*
	 * get the initial codebook and do more consistency checking
	 */
	fea_skiprec(init_strm, (long) init_rec - 1, init_ih);
	icbk = allo_vqfea_rec(init_ih);
	if (get_vqfea_rec(icbk, init_ih, init_strm) == EOF)    
	    ERROR_EXIT("EOF when trying to read initial codebook");
	if (*icbk->current_size > c_rows) 
	    ERROR_EXIT("initial codebook is too big");
	if (*icbk->current_size == c_rows && init_behav == INIT_NOCLUSTER) 
	    ERROR_EXIT(
	    "initial codebook size == design size with INIT_NOCLUSTER specified");
	if ((cbk_type != -1) && (*icbk->cbk_type != cbk_type)) {
	    Fprintf(histrm, 
		"vqdes: initial codebook type %s differs from param file value %s\n", 
		vq_cbk_types[*icbk->cbk_type], vq_cbk_types[cbk_type]);
	    Fprintf(stderr, 
		"vqdes: initial codebook type %s differs from param file value %s\n", 
		vq_cbk_types[*icbk->cbk_type], vq_cbk_types[cbk_type]);
	    exit(1);
	}
	if (*icbk->dimen != fea_dim) 
	    ERROR_EXIT("dimension in initial codebook doesn't match training input");
	/*
	 *check for generic header item giving FEA field
	 */
	vqfea_type = genhd_type("fea_field", (int *)NULL, init_ih);
	if (vqfea_type == HD_UNDEF || vqfea_type != CHAR) { 
	    Fprintf(histrm, 
	    "vqdes: WARNING - initial codebook has no generic header item fea_field\n");
	    Fprintf(stderr, 
	    "vqdes: WARNING - initial codebook has no generic header item fea_field\n");
	}
	if (strcmp(get_genhd("fea_field", init_ih), fea_field) != 0) {
	    Fprintf(histrm, 
		"vqdes: WARNING - inconsistent fea_field in initial codebook\n");
	    Fprintf(histrm, "\tfield from parameter file: %s\n", fea_field);
	    Fprintf(histrm, "\tfield from initial codebook: %s\n", get_genhd("fea_field",init_ih));
	    Fprintf(stderr, 
		"vqdes: WARNING - inconsistent fea_field in initial codebook\n");
	    Fprintf(stderr, "\tfield from parameter file: %s\n", fea_field);
	    Fprintf(stderr, "\tfield from initial codebook: %s\n", get_genhd("fea_field",init_ih));
	}
	/* Now copy the initial codebook
	 */
	*cbk->current_size = *icbk->current_size;
	for (i = 0; i < *cbk->current_size; i++) 
	    for (j = 0; j < fea_dim; j++) 
		cbk->codebook[i][j] = icbk->codebook[i][j];
    }
    else {
	/* no initial codebook -- start from scratch
 	 */	
	*cbk->current_size = 0;
	init_behav = INIT_NOCLUSTER;
    }
    /*These parameters apply whether or not there's an initial codebook.
     */
    *cbk->num_iter = 0;
    *cbk->dimen = fea_dim;
    *cbk->cbk_struct = fea_encode(getsym_s("cbk_struct"), "cbk_struct", vq_oh);
    if (*cbk->cbk_struct == -1) s_param_err(getsym_s("cbk_struct"));
    *cbk->design_size = c_rows;
    *cbk->conv_ratio = getsym_d("conv_ratio");
    *cbk->cbk_type = cbk_type;
    *cbk->dist_type = fea_encode(getsym_s("dist_type"), "dist_type", vq_oh);
    if (*cbk->dist_type == -1) s_param_err(getsym_s("dist_type"));    
    symerr_exit();  /*exit if any of the parameters were missing*/
/*
 *Create and write header for checkpoint file.
 */
    chk_oh = copy_header(vq_oh);
    TRYOPEN(argv[0], checkfile, "w", chkstrm);
    (void) strcpy (chk_oh->common.prog, "vqdes");
    (void) strcpy (chk_oh->common.vers, Version);
    (void) strcpy (chk_oh->common.progdate, Date);
    add_source_file(chk_oh, input_ana, fea_ih);
    if (init_flag) add_source_file(chk_oh, init_file, init_ih);
    add_comment(chk_oh, cmd_line);
    (void) add_comment(chk_oh, "This is a checkpoint file.");
    write_header(chk_oh, chkstrm);    	
/*
 *Allocate and do pre-processing of FEA data array
 */
    if (ndrec_ana != -1 && ndrec_ana * fea_dim <= FEA_BUFSIZE)
	/* number of FEAs is known, and all will fit */
    {
	whole_fits = 1;
	fea_len = ndrec_ana;
    }
    else /* unknown number, or too many to process at once */
    {
	whole_fits = 0;
	fea_len = FEA_BUFSIZE / fea_dim;
    }
    feadata = f_mat_alloc((unsigned) fea_len, (unsigned) fea_dim);
    if (feadata == NULL) ERROR_EXIT("couldn't allocate enough memory");
    if (whole_fits) {
	/*go ahead and fill the array now*/ 
	fea_ptr = (float *) get_fea_ptr(fea_rec, fea_field, fea_ih);
	for (i = 0; i < ndrec_ana; i++) {
	    (void) get_fea_rec(fea_rec, fea_ih, fea_strm);
	    vec_copy_f(feadata[i], fea_ptr, fea_dim);
	}
    }
/*
 *initialize the space for the code points returned by vqdesign
 */
    enc = (long *) calloc((unsigned) fea_len, sizeof(long));    
/*
 *select distortion function
 */
    switch (*cbk->dist_type) {
	case MSE:  
	    distort = NULL;
	    break;
	case MSE_LAR:
	    distort = mse_lar_dist;
	    break;
	default:
	    ERROR_EXIT("invalid distortion type");
    }
/*
 * history output of design parameters
 */
    if (debug_level > 0) {
        Fprintf(histrm, "\nParameter values:\n");
	Fprintf(histrm, "\tdimension of training vectors = %d\n", fea_dim);
	Fprintf(histrm, "\tname of FEA field with training vectors = %s\n", fea_field);
	Fprintf(histrm, "\tfractional conv. thresh = %g\n", *cbk->conv_ratio);
	Fprintf(histrm, "\tmax number of iterations on one level = %d\n",
	    max_iter);
	Fprintf(histrm, "\tsize of VQ codebook= %d\n", *cbk->design_size);
	Fprintf(histrm, "\tdistortion type = %s\n", dist_types[*cbk->dist_type]);
	Fprintf(histrm, "\tcodebook structure = %s\n", cbk_structs[*cbk->cbk_struct]);
	if (init_flag) Fprintf(histrm, 
	    "Initial codebook supplied from file %s, record %d\n", init_file, init_rec);
    }	
/*
 * main processing goes here
 */
    switch (*cbk->cbk_struct) {
	case FULL_SEARCH:  
	    vq_ret = vqdesign(histrm, debug_level, feadata, fea_len, fea_dim, 
		cbk, enc, init_behav, SPLIT_DIST, rc_get_chunk, 
		(int (*)())NULL, distort, (int (*)())NULL, checkpnt, max_iter);
	    if (vq_ret == VQ_NOCONVG) Fprintf(stderr, 
		"vqdes: FAILED - codebook didn't converge, see history file\n");
    if (debug_level > 0) Fprintf(histrm, 
		"vqdes: vqdesign return %s\n", vq_returns[vq_ret]);
	    break;
	default:
	    ERROR_EXIT("invalid distortion type");
    }
/*
 *write FEA record for final codebook
 */
    put_vqfea_rec(cbk, vq_oh, vq_strm);

    exit(0);
}

void
s_param_err(string)
char *string;
{
    Fprintf(stderr, "vqdes: unknown parameter string %s\n", string);
    exit (1);
}

long
rc_get_chunk(data, len, dim, num_prev, error)
float **data;
long len;
long dim;
long num_prev;
int *error;
{
/*get_chunk routine (for vqdesign) to get FEA training data
 */
    long    i, lim;
    float   *fea_ptr;

    *error = 0;
    if (whole_fits) {
	if (num_prev == 0) 
	    return len;
	else
	    return 0;
    }
    /*we get here only if the whole FEA file didn't fit in data
     */
    if (num_prev == ndrec_ana)
	return 0;

    if (num_prev == 0) {
	rewind(fea_strm);
	fea_ih = read_header(fea_strm);
    }

    fea_ptr = (float *) get_fea_ptr(fea_rec, fea_field, fea_ih);

    /* If the number of remaining input records is unknown or
     * greater than "len", try to fill "data" completely;
     * otherwise try for the known number of remaining records.
     */
    lim = (ndrec_ana == -1 || ndrec_ana - num_prev > len)
	? len : ndrec_ana - num_prev;

    for (i = 0; i < lim; i++) {
	if (get_fea_rec(fea_rec, fea_ih, fea_strm) == EOF) {
	    if (ndrec_ana != -1)
		*error = 1;
	    break;
	}
	vec_copy_f(data[i], fea_ptr, dim);
    }

    return i;
}

int
checkpnt(cdbk, chk_type)
struct vqcbk *cdbk;		/*codebook pointer*/
int chk_type;			/*type of checkpoint*/
{
/*this checkpoint routine writes the codebook to the output file 
 *right before each split operation and also to the checkpoint file
 *right after every encode operation (pass through the training sequence).
 *Only one record is kept in the checkpoint file.  
 */
    if (chk_type == CHK_SPLIT) {
	put_vqfea_rec(cdbk, vq_oh, vq_strm);
	Fflush(vq_strm);
    }
    rewind(chkstrm);
    write_header(chk_oh, chkstrm);
    put_vqfea_rec(cdbk, chk_oh, chkstrm);
    Fflush(chkstrm);
}

#define LOG10 2.302585093

double
mse_lar_dist(fea_vector, vec_dim, codeword, cdwd_dim, error)
float *fea_vector;
long vec_dim;
float *codeword;
long cdwd_dim;
int *error;
/*
 *This distortion function assumes both the feature vector and 
 *the codeword are reflection coefficients; it converts both to 
 *log area ratios and then does mean-square error.  This program is 
 *quite inefficient - if we really want to design large codebooks 
 *using mse_lar, training sequence should be converted to LARs 
 *before encoding.  
 */
{
    double diff;
    float fea_lar, cd_lar;
    double msedist = 0.0;
    double temp;
    while (vec_dim-- != 0) {
	temp = *fea_vector++;
	fea_lar = log((1.0 + temp) / (1.0 - temp));
	temp = *codeword++;
	cd_lar = log((1.0 + temp) / (1.0 - temp));
	diff = fea_lar - cd_lar;
	msedist += diff * diff;
    }
    *error = 0;
    return msedist / cdwd_dim;
}    

void
vec_copy_f(dst, src, n)
float  *dst, *src;	/*pointers to destination and source vectors*/
long n;		/*size of vectors*/
/*
 *This routine copies src to dst.
 */
{
    while (n-- != 0)
	*dst++ = *src++;
}
